### DISCUSSION TITLE
Support writing arbitrary key/value pairs to container metadata

### DISCUSSION THREAD

**Discussion by Michael-Equi** [PyAV-Org/PyAV#1601] [2. Feature Ideas]
## Desired Behavior

Currently only the title is allowed to be written to a containers metadata. The behavior is tested in the following unit test: https://github.com/PyAV-Org/PyAV/blob/86f66ac1c6682c5c07aeb8df4e070b0fe3723fb7/tests/test_encode.py#L74 

I have confirmed ffmpeg does support this by using the following CLI command given the path to a libx264 encoded video.

```
subprocess.run(
        [
            "ffmpeg",
            "-y",  # Overwrite output files without asking
            "-i",
            input,
            "-c",
            "copy",
            "-copyts",  # Copy timestamps from input to output
            "-metadata",
            f"key={value}",
            "-movflags",
            "+use_metadata_tags",
            output,
        ],
``` 

## Example API

Allow any arbitrary key/value pair to be written to the metadata in the same way title can be written to the metadata.

**Comment by dotysan** [comment]
In addition to adding metadata to the container. Let's also be able to add it to the streams.
 - `-metadata:s:v:0 source=ACMEcam`
 - `-metadata:s:a:0 sound=Silence`

**Comment by dogganon** [comment]
~~This problem seems to be specific to the mp4 implementation. Switching to mkv allows any metadata to be set on both the container and the containing streams.~~

~~@WyattBlue This should probably be reported as a BUG.~~

**Edit:**

Problem solved after passing `use_metadata_tags` when writing the mp4 file:
```python
av.open(resource, mode="w", format="mp4", options={"movflags": "+use_metadata_tags"})
```

@WyattBlue Maybe it's worth considering adding [examples](https://github.com/PyAV-Org/PyAV/tree/main/examples) on how to work with metadata.

**Comment by WyattBlue** [comment]
(newer) PyAV already supports arbitrary container metadata. For MP4, FFmpeg requires the use_metadata_tags muxer flag for nonstandard keys:

```python
with av.open(
    output, mode="w", format="mp4", options={"movflags": "+use_metadata_tags"},
) as container:
    container.metadata["key"] = "value"
```

Make sure metadata is set before encoding or muxing begins, since the metadata is written with the container header.

Streams also expose a metadata dictionary:

```python
stream = container.add_stream("libx264", rate=30)
stream.metadata["source"] = "ACMEcam"
```

Arbitrary stream metadata works with formats such as Matroska. MP4 only preserves certain stream-level metadata fields, even when use_metadata_tags is enabled, which is a limitation of FFmpeg’s MP4 muxer rather than PyAV.
